internal/upload/container/container.go

```diff
@@ -0,0 +1,183 @@
+package container
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	_ "github.com/containers/image/v5/docker/archive"
+	_ "github.com/containers/image/v5/oci/archive"
+	_ "github.com/containers/image/v5/oci/layout"
+
+	"github.com/containers/common/pkg/retry"
+	"github.com/containers/image/v5/copy"
+	"github.com/containers/image/v5/docker"
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/manifest"
+	"github.com/containers/image/v5/signature"
+	"github.com/containers/image/v5/transports"
+	"github.com/containers/image/v5/types"
+	"github.com/opencontainers/go-digest"
+)
+
+const (
+	DefaultUserAgent  = "osbuild-composer/1.0"
+	DefaultPolicyPath = "/etc/containers/policy.json"
+)
+
+type Credentials struct {
+	Username string
+	Password string
+}
+
+type Client struct {
+	Target reference.Named
+	Auth   Credentials
+
+	ReportWriter io.Writer
+
+	PrecomputeDigests bool
+	MaxRetries        int
+
+	UserAgent string
+	TlsVerify bool
+
+	policy *signature.Policy
+	sysCtx *types.SystemContext
+}
+
+func NewClient(destination string) (*Client, error) {
```

Comment:
From the usage, it seems that the `username` and `password` are mandatory. If so, I think the `NewClient` method should receive them instead of having the caller set them separately

Reply:
The reason it is not is because I plan to add this hybrid support so that in case they are not explicitly set we can read them from a config file, i.e. in case of brew/koji.

Reply:
If so, then shouldn't the `username` and `password` on the `ContainerTargetOptions` be optional? Or am I missing something?

Reply:
With optional you mean they should be pointer-to-strings?

Reply:
In its essence I mean that the caller does not have to pass them in the API (with the hybrid support). I think that it does translate to being a pointer-to-string in the Go struct and define it as `omitempty` (same as you do for `TlsVerify`)